catalog/core/catalog-core-downloadaction/src/main/java/org/codice/ddf/catalog/download/action/ResourceDownloadActionProvider.java

```diff
@@ -0,0 +1,145 @@
+import org.codice.ddf.configuration.SystemBaseUrl;
+import org.codice.ddf.configuration.SystemInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.action.Action;
+import ddf.action.impl.ActionImpl;
+import ddf.catalog.data.Metacard;
+
+/**
+ * Action provider that creates {@link Action}s used to asynchronously download resources to
+ * the product cache.
+ */
+public class ResourceDownloadActionProvider extends AbstractMetacardActionProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceDownloadActionProvider.class);
+    
+    private static final String TITLE = "Copy resource to local site";
+
+    private static final String DESCRIPTION =
+            "Copies the resource associated with this metacard to the local site";
+
+    private static final String LOCAL_CONTENT_SCHEME = "content:";
+
+    private final ResourceCacheServiceMBean resourceCacheMBean;
+    
+    public ResourceDownloadActionProvider(String actionProviderId) {
+        super(actionProviderId, TITLE, DESCRIPTION);
+        resourceCacheMBean = createResourceCacheMBeanProxy();
+    }
+
+    /**
+     * Returns {@code true} if the resource associated with the {@link Metacard} is local; false otherwise.
+     * <p>
+     * {@inheritDoc}
+     */
+    @Override
+    protected boolean canHandleMetacard(Metacard metacard) {
+        return hasResourceUri(metacard)
+                ? (hasRemoteSiteName(metacard) || isResourceUriRemote(metacard)) && !isResourceCached(metacard) : false;
+    }
+    
+    @Override
+    protected Action createMetacardAction(String actionProviderId, String title, String description,
+            URL url) {
+        return new ActionImpl(actionProviderId, title, description, url);
+    }
+
+    @Override
+    protected URL getMetacardActionUrl(String metacardSource, Metacard metacard) throws Exception {
+        String encodedMetacardId = URLEncoder.encode(metacard.getId(), CharEncoding.UTF_8);
+        String encodedMetacardSource = URLEncoder.encode(metacardSource, CharEncoding.UTF_8);
+        return getActionUrl(encodedMetacardSource, encodedMetacardId);
+    }
+    
+    private URL getActionUrl(String metacardSource, String metacardId) throws Exception {
+        return new URI(SystemBaseUrl.constructUrl(String.format("%s?%s=%s&%s=%s",
+                CONTEXT_PATH,
+                SOURCE_PARAM,
+                metacardSource,
+                METACARD_PARAM,
+                metacardId), true)).toURL();
+    }
+    
+    private boolean isResourceCached(Metacard metacard) {
+        return resourceCacheMBean.contains(metacard);
+    }
+
+    private boolean hasResourceUri(Metacard metacard) {
+        Optional<String> resourceUri = Optional.ofNullable(metacard.getResourceURI())
+                .map(uri -> uri.toString());
+        return resourceUri.isPresent();
+    }
+
+    private boolean isResourceUriRemote(Metacard metacard) {
```

Comment:
Not off the top of my head.  

I prefer to move things into a common place once there are two or three places that need it.  I do not like to prematurely extract code we think might be common before we are sure.  That sometimes leads to the wrong or messy abstractions.